### ISSUE TITLE
Unable to customize Undertow handlers through UndertowBuilderCustomizer

### ISSUE THREAD

**Issue by tollrakesh** [spring-projects/spring-boot#4374]
Hi,

I would want to block certain IP addresses from accessing my application. For doing this, when Undertow is used, the recommended way is to use custom handlers. So, to add a custom handler, I need to use `UndertowBuilderCustomizer`. 

But, when a handler is added this way, `UndertowEmbeddedServletContainer` doesnt seem to check the existence of a handler from the customizer. Due to this, though I add a handler (`IPAddressAccessControlHandler` in my case), it isnt working correctly.

**Comment by snicoll** [comment]
Can you show what you've tried? Or a reference to what Undertow recommends?

**Comment by tollrakesh** [comment]
Below is the snippet

        @Bean
        public UndertowEmbeddedServletContainerFactory embeddedServletContainerFactory() {
            UndertowEmbeddedServletContainerFactory factory = new UndertowEmbeddedServletContainerFactory();
            factory.addBuilderCustomizers(new UndertowBuilderCustomizer() {

                @Override
                public void customize(Undertow.Builder builder) {
                    builder.setHandler(accessControlHandler());
                }

                public IPAddressAccessControlHandler accessControlHandler() {
                    IPAddressAccessControlHandler accessControlHandler = new IPAddressAccessControlHandler();
                    accessControlHandler.addDeny("x.x.x.x");
                    return accessControlHandler;
                }

            });
            return factory;
        }

Basically, I'm looking for a way to "add" a new handler, without sub-classing the original code from Spring Boot.

I understand that the Undertow.Builder only has a set method. But, since there is a factory class already made available by Spring Boot, I think we just need to be able to add additional handler chain as needed to the builder customizer and have the factory add the additional handler chain as needed.

**Comment by snicoll** [comment]
Thank you, that helps a lot.

**Comment by benc** [comment]
There is a `addOuterHandlerChainWrapper` feature of the deployment customizer that already does that.

This code works for me:

```
@Bean
public UndertowEmbeddedServletContainerFactory embeddedServletContainerFactory() {
    UndertowEmbeddedServletContainerFactory factory = new UndertowEmbeddedServletContainerFactory();

    factory.addDeploymentInfoCustomizers(deploymentInfo -> {
        deploymentInfo.addOuterHandlerChainWrapper(CustomHandler::new);
    });

    return factory;
}
```

**Comment by wilkinsona** [comment]
@benc Thank you

@tollrakesh There are also `addInnerHandlerChainWrapper` that may be of interest. As far as I can tell, one of other of these methods will meet your needs. Please let us know if that's not the case.

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by spring-issuemaster** [comment]
Closing due to lack of requested feedback. If you would like us to look at this issue, please provide the requested information and we will re-open the issue.

**Comment by mindiga** [comment]
I had the same issue and resolved it through

	@Bean
	public UndertowServletWebServerFactory embeddedServletContainerFactory()
	{
		UndertowServletWebServerFactory factory = new UndertowServletWebServerFactory();
		factory.addDeploymentInfoCustomizers(new UndertowDeploymentInfoCustomizer() {
			@Override
			public void customize(DeploymentInfo deploymentInfo)
			{
				deploymentInfo.addInitialHandlerChainWrapper(MockupServerHttpHandler::new);
			}
		});
		
		return factory;
	}

apparently you have to use UndertowDeploymentInfoCustomizer instead of UndertowBuilderCustomizer
